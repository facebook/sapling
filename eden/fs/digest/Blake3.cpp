/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

#include "eden/fs/digest/Blake3.h"

#include <stdexcept>

#include <blake3.h> // @manual

namespace facebook::eden {

Blake3::Blake3() {
  blake3_hasher_init(&hasher_);
}

Blake3::Blake3(folly::ByteRange key) {
  if (key.size() != BLAKE3_KEY_LEN) {
    throw std::invalid_argument("Invalid key size, it must be 32 bytes length");
  }

  auto* const keyPtr = key.data();
  const uint8_t(&keyArray)[BLAKE3_KEY_LEN] =
      *reinterpret_cast<const uint8_t(*)[BLAKE3_KEY_LEN]>(keyPtr);
  blake3_hasher_init_keyed(&hasher_, keyArray);
}

void Blake3::update(const void* data, size_t size) {
  blake3_hasher_update(&hasher_, data, size);
}

void Blake3::update(folly::ByteRange data) {
  update(data.data(), data.size());
}

void Blake3::update(folly::StringPiece data) {
  update(data.data(), data.size());
}

void Blake3::finalize(folly::MutableByteRange out) {
  if (out.size() != BLAKE3_OUT_LEN) {
    throw std::invalid_argument("Unexpected len");
  }

  blake3_hasher_finalize(&hasher_, out.data(), BLAKE3_OUT_LEN);
}

} // namespace facebook::eden
