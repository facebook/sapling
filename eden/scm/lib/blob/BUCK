load("@fbcode_macros//build_defs:rust_library.bzl", "rust_library")
load("@fbcode_macros//build_defs/lib:rust_oss.bzl", "rust_oss")

oncall("sapling")

rust_library(
    name = "blob",
    srcs = glob(["src/**/*.rs"]),
    autocargo = {"cargo_toml_config": {
        "lib": {"name": "blob"},
        "package": {
            "authors": ["Meta Source Control Team <sourcecontrol-dev@meta.com>"],
            "homepage": "https://sapling-scm.com/",
            "license": "MIT",
            "name": "blob",
            "repository": "https://github.com/facebook/sapling",
        },
    }},
    named_deps = {
        "blake3-constants": "//blake3:blake3-constants-rust-nothrift",
    },
    deps = [
        "fbsource//third-party/rust:blake3",
        "fbsource//third-party/rust:bytes",
        "fbsource//third-party/rust:sha1",
        "//common/rust/folly:iobuf",
        "//eden/scm/lib/minibytes:minibytes",
        "//eden/scm/lib/types:types",
    ] + ([] if rust_oss.is_oss_build() else [
        "//common/rust/folly:iobuf",
    ]),
)
