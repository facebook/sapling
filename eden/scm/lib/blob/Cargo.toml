# @generated by autocargo from //eden/scm/lib/blob:scm-blob

[package]
name = "scm-blob"
version = "0.1.0"
authors = ["Meta Source Control Team <sourcecontrol-dev@meta.com>"]
edition = "2021"
homepage = "https://sapling-scm.com/"
repository = "https://github.com/facebook/sapling"
license = "MIT"

[dependencies]
blake3 = { version = "=1.5.5", features = ["mmap", "rayon", "traits-preview"] }
bytes = { version = "1.9.0", features = ["serde"] }
sapling-minibytes = { version = "0.1.0", path = "../minibytes" }
sapling-types = { version = "0.1.0", path = "../types" }
sha1 = "0.10.5"
