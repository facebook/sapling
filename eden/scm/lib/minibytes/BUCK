load("@fbcode_macros//build_defs:rust_library.bzl", "rust_library")

oncall("sapling")

rust_library(
    name = "minibytes",
    srcs = glob(["src/**/*.rs"]),
    autocargo = {"cargo_toml_config": {
        "features": {
            "default": [
                "frombytes",
                "frommmap",
                "non-zerocopy-into",
            ],
            "frombytes": ["bytes"],
            "frommmap": ["memmap2"],
            "non-zerocopy-into": [],
        },
        "lib": {"name": "minibytes"},
        "package": {
            "authors": ["Meta Source Control Team <sourcecontrol-dev@meta.com>"],
            "description": "Shared reference-counted bytes and str with zero-copy slicing support.",
            "homepage": "https://sapling-scm.com/",
            "license": "MIT",
            "name": "sapling-minibytes",
            "repository": "https://github.com/facebook/sapling",
        },
    }},
    crate_root = "src/lib.rs",
    features = [
        "frombytes",
        "frommmap",
        "non-zerocopy-into",
    ],
    test_deps = [
        "fbsource//third-party/rust:quickcheck",
        "fbsource//third-party/rust:serde_cbor",
    ],
    # A test inside this target is using #[should_panic], setting the backtrace
    # to false here, otherwise the test binary will try to extract the backtrace
    # and the test will fail with leak sanitizer.
    test_env = {"RUST_BACKTRACE": "0"},
    test_labels = ["tpx-rust-no-backtrace-workaround"],
    deps = [
        "fbsource//third-party/rust:bytes",
        "fbsource//third-party/rust:memmap2",
        "fbsource//third-party/rust:serde",
    ],
)
