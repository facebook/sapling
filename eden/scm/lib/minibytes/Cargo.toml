# @generated by autocargo from //eden/scm/lib/minibytes:minibytes

[package]
name = "sapling-minibytes"
version = "0.1.0"
authors = ["Meta Source Control Team <sourcecontrol-dev@meta.com>"]
edition = "2021"
description = "Shared reference-counted bytes and str with zero-copy slicing support."
homepage = "https://sapling-scm.com/"
repository = "https://github.com/facebook/sapling"
license = "MIT"

[lib]
name = "minibytes"

[dependencies]
bytes = { version = "1.9.0", features = ["serde"], optional = true }
memmap2 = { version = "0.9.5", optional = true }
serde = { version = "1.0.185", features = ["derive", "rc"] }

[dev-dependencies]
quickcheck = "1.0"
serde_cbor = "0.11"

[features]
default = ["frombytes", "frommmap", "non-zerocopy-into"]
frombytes = ["bytes"]
frommmap = ["memmap2"]
non-zerocopy-into = []
