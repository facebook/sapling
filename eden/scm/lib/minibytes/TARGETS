load("@fbcode_macros//build_defs:rust_library.bzl", "rust_library")
load("@fbsource//tools/build_defs:glob_defs.bzl", "glob")

oncall("mercurial_release")

rust_library(
    name = "minibytes",
    srcs = glob(["src/**/*.rs"]),
    autocargo = {
        "cargo_toml_config": {
            "features": {
                "default": [
                    "frombytes",
                    "frommmap",
                ],
                "frombytes": ["bytes"],
                "frommmap": ["memmap2"],
            },
            "package": {
                "license": "MIT",
            },
        },
    },
    crate_root = "src/lib.rs",
    features = [
        "frombytes",
        "frommmap",
    ],
    test_deps = [
        "fbsource//third-party/rust:quickcheck",
    ],
    # A test inside this target is using #[should_panic], setting the backtrace
    # to false here, otherwise the test binary will try to extract the backtrace
    # and the test will fail with leak sanatizer.
    test_env = {"RUST_BACKTRACE": "0"},
    test_labels = ["tpx-rust-no-backtrace-workaround"],
    deps = [
        "fbsource//third-party/rust:bytes",
        "fbsource//third-party/rust:memmap2",
        "fbsource//third-party/rust:serde",
    ],
)
