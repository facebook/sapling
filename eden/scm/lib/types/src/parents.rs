/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use serde_derive::Deserialize;
use serde_derive::Serialize;

use crate::Key;
use crate::hgid::HgId;
use crate::hgid::NULL_ID;

/// Enum representing a Mercurial hgid's parents.
///
/// A hgid may have zero, one, or two parents (referred to as p1 and p2 respectively).
/// Ordinarily, a non-existent parent is denoted by a null hash, consisting of all zeros.
/// A null p1 implies a null p2, so it is invalid for a hgid to have a p2 without a p1.
///
/// In Rust, these restrictions can be enforced with an enum that makes invalid
/// states unrepresentable.
#[derive(
    Copy,
    Clone,
    Debug,
    Eq,
    Hash,
    Ord,
    PartialEq,
    PartialOrd,
    Serialize,
    Deserialize
)]
#[cfg_attr(
    any(test, feature = "for-tests"),
    derive(quickcheck_arbitrary_derive::Arbitrary)
)]
#[serde(untagged)]
#[derive(Default)]
pub enum Parents {
    #[default]
    None,
    One(HgId),
    Two(HgId, HgId),
}

impl Parents {
    /// Construct a new Parents from two potentially null HgId hashes.
    /// This function will panic if an invalid combination of Nodes is given --
    /// namely, if p1 is null but p2 is not null.
    pub fn new(p1: HgId, p2: HgId) -> Self {
        match (p1.is_null(), p2.is_null()) {
            (true, true) => Parents::None,
            (false, true) => Parents::One(p1),
            (true, _) => panic!("invalid parents: non-null p2 with null p1"),
            (false, false) => Parents::Two(p1, p2),
        }
    }

    /// Convert this Parents into a tuple representation, with non-existent
    /// parents represented by NULL_ID.
    pub fn into_nodes(self) -> (HgId, HgId) {
        match self {
            Parents::None => (NULL_ID, NULL_ID),
            Parents::One(p1) => (p1, NULL_ID),
            Parents::Two(p1, p2) => (p1, p2),
        }
    }

    pub fn p1(&self) -> Option<&HgId> {
        match self {
            Parents::None => None,
            Parents::One(p1) => Some(p1),
            Parents::Two(p1, _) => Some(p1),
        }
    }

    pub fn p2(&self) -> Option<&HgId> {
        match self {
            Parents::None | Parents::One(_) => None,
            Parents::Two(_, p2) => Some(p2),
        }
    }

    pub fn to_keys(&self) -> [Key; 2] {
        let (p1, p2) = self.into_nodes();
        [
            Key::new(Default::default(), p1),
            Key::new(Default::default(), p2),
        ]
    }

    /// Reports `Vec<HgId>` excluding `NULL_ID`s. Might reutrn an empty vec.
    pub fn to_vec(&self) -> Vec<HgId> {
        let mut result = Vec::new();
        match self {
            Parents::None => {}
            Parents::One(p1) => {
                if !p1.is_null() {
                    result.push(*p1);
                }
            }
            Parents::Two(p1, p2) => {
                if !p1.is_null() {
                    result.push(*p1);
                }
                if !p2.is_null() {
                    result.push(*p2);
                }
            }
        }
        result
    }
}

impl FromIterator<HgId> for Parents {
    fn from_iter<I: IntoIterator<Item = HgId>>(iter: I) -> Self {
        let mut iter = iter.into_iter();
        let p1 = iter.next().unwrap_or(NULL_ID);
        let p2 = iter.next().unwrap_or(NULL_ID);
        Parents::new(p1, p2)
    }
}

impl IntoIterator for Parents {
    type IntoIter = ParentIter;
    type Item = HgId;

    fn into_iter(self) -> ParentIter {
        ParentIter(self)
    }
}

impl IntoIterator for &Parents {
    type IntoIter = ParentIter;
    type Item = HgId;

    fn into_iter(self) -> ParentIter {
        ParentIter(self.clone())
    }
}

#[derive(Debug)]
pub struct ParentIter(Parents);

impl Iterator for ParentIter {
    type Item = HgId;

    fn next(&mut self) -> Option<Self::Item> {
        match self.0 {
            Parents::None => None,
            Parents::One(p1) => {
                self.0 = Parents::None;
                Some(p1)
            }
            Parents::Two(p1, p2) => {
                self.0 = Parents::One(p2);
                Some(p1)
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use serde_json;
    use serde_json::json;

    use super::*;

    #[test]
    fn from_iter() {
        let none = Parents::from_iter(vec![NULL_ID, NULL_ID]);
        assert_eq!(none, Parents::None);

        let p1 = HgId::from_byte_array([0xAA; 20]);
        let one = Parents::from_iter(vec![p1, NULL_ID]);
        assert_eq!(one, Parents::One(p1));

        let p2 = HgId::from_byte_array([0xBB; 20]);
        let two = Parents::from_iter(vec![p1, p2]);
        assert_eq!(two, Parents::Two(p1, p2));
    }

    #[test]
    #[should_panic]
    fn from_iter_invalid() {
        let p2 = HgId::from_byte_array([0xAA; 20]);
        let _ = Parents::from_iter(vec![NULL_ID, p2]);
    }

    #[test]
    fn into_iter() {
        let parents = Parents::None;
        let none = parents.into_iter().collect::<Vec<_>>();
        assert_eq!(none, Vec::new());

        let p1 = HgId::from_byte_array([0xAA; 20]);
        let parents = Parents::One(p1);
        let one = parents.into_iter().collect::<Vec<_>>();
        assert_eq!(one, vec![p1]);

        let p2 = HgId::from_byte_array([0xBB; 20]);
        let parents = Parents::Two(p1, p2);
        let two = parents.into_iter().collect::<Vec<_>>();
        assert_eq!(two, vec![p1, p2]);
    }

    #[test]
    fn untagged_serialization() {
        let parents = Parents::None;
        let none = serde_json::to_value(parents).unwrap();
        assert_eq!(none, json!(null));

        let p1 = HgId::from_byte_array([0x1; 20]);
        let parents = Parents::One(p1);
        let one = serde_json::to_value(parents).unwrap();
        let expected = json!([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]);
        assert_eq!(one, expected);

        let p2 = HgId::from_byte_array([0x2; 20]);
        let parents = Parents::Two(p1, p2);
        let two = serde_json::to_value(parents).unwrap();
        let p1_json = json!([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]);
        let p2_json = json!([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]);
        assert_eq!(two, json!([p1_json, p2_json]));
    }

    #[test]
    fn serialized_size_cbor() {
        let parents = Parents::None;
        let none = serde_cbor::to_vec(&parents).unwrap();
        assert_eq!(none.len(), 1);

        let p1 = HgId::from_byte_array([0x1; 20]);
        let parents = Parents::One(p1);
        let one = serde_cbor::to_vec(&parents).unwrap();
        assert_eq!(one.len(), 21);

        let p2 = HgId::from_byte_array([0x2; 20]);
        let parents = Parents::Two(p1, p2);
        let two = serde_cbor::to_vec(&parents).unwrap();
        assert_eq!(two.len(), 43);
    }
}
