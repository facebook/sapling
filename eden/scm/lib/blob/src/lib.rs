/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#![allow(unexpected_cfgs)]

use bytes::Buf;
use types::Blake3;
use types::Sha1;

#[derive(Clone, Debug)]
pub enum Blob {
    Bytes(minibytes::Bytes),
    #[cfg(fbcode_build)]
    IOBuf(iobuf::IOBufShared),
}

impl Blob {
    pub fn to_bytes(&self) -> minibytes::Bytes {
        match self {
            Self::Bytes(bytes) => bytes.clone(),
            #[cfg(fbcode_build)]
            Self::IOBuf(buf) => minibytes::Bytes::from(Vec::<u8>::from(buf.clone())),
        }
    }

    pub fn into_bytes(self) -> minibytes::Bytes {
        match self {
            Self::Bytes(bytes) => bytes,
            #[cfg(fbcode_build)]
            Self::IOBuf(buf) => minibytes::Bytes::from(Vec::<u8>::from(buf)),
        }
    }

    pub fn into_vec(self) -> Vec<u8> {
        match self {
            Self::Bytes(bytes) => bytes.into(),
            #[cfg(fbcode_build)]
            Self::IOBuf(buf) => Vec::<u8>::from(buf),
        }
    }

    #[cfg(fbcode_build)]
    pub fn into_iobuf(self) -> iobuf::IOBufShared {
        match self {
            Self::Bytes(bytes) => iobuf::IOBufShared::from(bytes),
            Self::IOBuf(buf) => buf,
        }
    }

    pub fn len(&self) -> usize {
        match self {
            Self::Bytes(bytes) => bytes.len(),
            #[cfg(fbcode_build)]
            Self::IOBuf(buf) => buf.len(),
        }
    }

    pub fn is_empty(&self) -> bool {
        match self {
            Self::Bytes(bytes) => bytes.is_empty(),
            #[cfg(fbcode_build)]
            Self::IOBuf(buf) => buf.is_empty(),
        }
    }

    pub fn sha1(&self) -> types::Sha1 {
        use sha1::Digest;

        let mut hasher = sha1::Sha1::new();

        match self {
            Self::Bytes(bytes) => {
                hasher.update(bytes);
            }
            #[cfg(fbcode_build)]
            Self::IOBuf(buf) => {
                let mut cur = buf.clone().cursor();

                while cur.has_remaining() {
                    let b = cur.chunk();
                    hasher.update(b);
                    cur.advance(b.len());
                }
            }
        }

        let bytes: [u8; Sha1::len()] = hasher.finalize().into();
        Sha1::from(bytes)
    }

    pub fn blake3(&self) -> types::Blake3 {
        use blake3::Hasher;

        #[cfg(fbcode_build)]
        let key = blake3_constants::BLAKE3_HASH_KEY;
        #[cfg(not(fbcode_build))]
        let key = b"20220728-2357111317192329313741#";

        let mut hasher = Hasher::new_keyed(key);

        match self {
            Self::Bytes(bytes) => {
                hasher.update(bytes);
            }
            #[cfg(fbcode_build)]
            Self::IOBuf(buf) => {
                let mut cur = buf.clone().cursor();

                while cur.has_remaining() {
                    let b = cur.chunk();
                    hasher.update(b);
                    cur.advance(b.len());
                }
            }
        }

        let hashed_bytes: [u8; Blake3::len()] = hasher.finalize().into();
        Blake3::from(hashed_bytes)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[cfg(fbcode_build)]
    #[test]
    fn test_iobuf_sha1_and_blake3() {
        let blob1 = Blob::Bytes(minibytes::Bytes::from("hello world!"));

        let blob2 = {
            let mut iobuf = iobuf::IOBufShared::from("hello");
            iobuf.append_to_end(iobuf::IOBufShared::from(" world!"));
            Blob::IOBuf(iobuf)
        };

        assert_eq!(blob1.sha1(), blob2.sha1());
        assert_eq!(blob1.blake3(), blob2.blake3());
    }
}
