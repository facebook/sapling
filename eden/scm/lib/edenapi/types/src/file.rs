/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use blob::Blob;
use minibytes::Bytes;
#[cfg(any(test, feature = "for-tests"))]
use quickcheck::Arbitrary;
#[cfg(any(test, feature = "for-tests"))]
use quickcheck_arbitrary_derive::Arbitrary;
use revisionstore_types::Metadata;
use serde_derive::Deserialize;
use serde_derive::Serialize;
use thiserror::Error;
use type_macros::auto_wire;
use types::hgid::HgId;
use types::key::Key;
use types::parents::Parents;

use crate::Blake3;
use crate::InvalidHgId;
use crate::ServerError;
use crate::Sha1;
use crate::UploadToken;
use crate::hash::check_hash;

/// Tombstone string that replaces the content of redacted files.
/// TODO(T48685378): Handle redacted content in a less hacky way.
const REDACTED_TOMBSTONE: &str = "PoUOK1GkdH6Xtx5j9WKYew3dZXspyfkahcNkhV6MJ4rhyNICTvX0nxmbCImFoT0oH\
      AF9ivWGaC6ByswQZUgf1nlyxcDcahHknJS15Vl9Lvc4NokYhMg0mV1rapq1a4bhNo\
      UI9EWTBiAkYmkadkO3YQXV0TAjyhUQWxxLVskjOwiiFPdL1l1pdYYCLTE3CpgOoxQ\
      V3EPVxGUPh1FGfk7F9Myv22qN1sUPSNN4h3IFfm2NNPRFgWPDsqAcaQ7BUSKa\n";

#[derive(Debug, Error)]
pub enum FileError {
    #[error(transparent)]
    Corrupt(InvalidHgId),
    /// File entry was redacted by the server. The received content
    /// did not validate but matches the known tombstone content for
    /// redacted data.
    #[error("Content for {0} is redacted")]
    Redacted(Key, Bytes),
    #[error("Can't validate filenode hash for {0} because it is an LFS pointer")]
    Lfs(Key, Bytes),
    #[error("Content for {0} is unavailable")]
    MissingContent(Key),
}

impl FileError {
    /// Get the data anyway, despite the error.
    pub fn data(&self) -> Bytes {
        use FileError::*;
        match self {
            Corrupt(InvalidHgId { data, .. }) => data,
            Redacted(_, data) => data,
            Lfs(_, data) => data,
            MissingContent(_) => panic!("no content attribute available for FileEntry"),
        }
        .clone()
    }
}

/// File "aux data", requires an additional mononoke blobstore lookup. See mononoke_types::ContentMetadataV2.
#[auto_wire]
#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize)]
pub struct FileAuxData {
    #[id(0)]
    pub total_size: u64,
    // #[id(1)] # deprecated
    #[id(2)]
    pub sha1: Sha1,
    // #[id(3)] # deprecated
    #[id(4)]
    pub blake3: Blake3,
    // None 'file_header_metadata' would mean file_header_metadata is not fetched/not known if it is present
    // Empty metadata would be translated into empty blob
    #[id(5)]
    pub file_header_metadata: Option<Bytes>,
}

/// File content
#[derive(Clone, Debug, Default, Deserialize, Serialize, Eq, PartialEq)]
pub struct FileContent {
    pub hg_file_blob: Bytes,
    pub metadata: Metadata,
}

impl FileContent {
    /// Get this entry's data. Checks data integrity but allows hash mismatches
    /// if the content is redacted or contains an LFS pointer.
    pub fn data(&self, key: &Key, parents: Parents) -> Result<Bytes, FileError> {
        use FileError::*;
        self.data_checked(key, parents).or_else(|e| match e {
            Corrupt(_) => Err(e),
            Redacted(..) => Ok(e.data()),
            Lfs(..) => Ok(e.data()),
            MissingContent(_) => Err(e),
        })
    }

    /// Get this entry's data after verifying the hgid hash.
    ///
    /// This method will return an error if the computed hash doesn't match
    /// the provided hash, regardless of the reason. Such mismatches are
    /// sometimes expected (e.g., for redacted files or LFS pointers), so most
    /// application logic should call `FileEntry::data` instead, which allows
    /// these exceptions. `FileEntry::data_checked` should only be used when
    /// strict filenode validation is required.
    pub fn data_checked(&self, key: &Key, parents: Parents) -> Result<Bytes, FileError> {
        // TODO(meyer): Clean this up, make LFS Pointers and redaction strongly typed all the way from here through scmstore
        let data = &self.hg_file_blob;

        // TODO(T48685378): Handle redacted content in a less hacky way.
        if data.len() == REDACTED_TOMBSTONE.len() && data == REDACTED_TOMBSTONE {
            return Err(FileError::Redacted(key.clone(), data.clone()));
        }

        // We can't check the hash of an LFS blob since it is computed using the
        // full file content, but the file entry only contains the LFS pointer.
        if self.metadata().is_lfs() {
            return Err(FileError::Lfs(key.clone(), data.clone()));
        }

        check_hash(data, parents, "blob", key.hgid).map_err(FileError::Corrupt)?;

        Ok(data.clone())
    }

    /// Get this entry's data without verifying the hgid hash.
    pub fn data_unchecked(&self) -> &Bytes {
        &self.hg_file_blob
    }

    /// Get this entry's metadata.
    pub fn metadata(&self) -> &Metadata {
        &self.metadata
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct FileResponse {
    pub key: Key,
    pub result: Result<FileEntry, ServerError>,
}

/// Structure representing source control file content on the wire.
/// Includes the information required to add the data to a mutable store,
/// along with the parents for hash validation.
#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct FileEntry {
    pub key: Key,
    pub parents: Parents,
    pub content: Option<FileContent>,
    pub aux_data: Option<FileAuxData>,
}

impl FileAuxData {
    /// Calculate `FileAuxData` from file content.
    pub fn from_content(blob: &Blob) -> Self {
        let total_size = blob.len() as _;
        Self {
            total_size,
            sha1: blob.sha1(),
            blake3: blob.blake3(),
            file_header_metadata: None, // can't be calculated
        }
    }
}

impl FileEntry {
    pub fn new(key: Key, parents: Parents) -> Self {
        Self {
            key,
            parents,
            ..Default::default()
        }
    }

    pub fn with_content(mut self, content: FileContent) -> Self {
        self.content = Some(content);
        self
    }

    pub fn with_aux_data(mut self, aux_data: FileAuxData) -> Self {
        self.aux_data = Some(aux_data);
        self
    }

    pub fn key(&self) -> &Key {
        &self.key
    }

    pub fn parents(&self) -> &Parents {
        &self.parents
    }

    pub fn aux_data(&self) -> Option<&FileAuxData> {
        self.aux_data.as_ref()
    }

    pub fn content(&self) -> Option<&FileContent> {
        self.content.as_ref()
    }

    pub fn data(&self) -> Result<Bytes, FileError> {
        self.content
            .as_ref()
            .ok_or_else(|| FileError::MissingContent(self.key().clone()))?
            .data(self.key(), *self.parents())
    }

    pub fn metadata(&self) -> Result<&Metadata, FileError> {
        Ok(self
            .content
            .as_ref()
            .ok_or_else(|| FileError::MissingContent(self.key().clone()))?
            .metadata())
    }
}

#[auto_wire]
#[derive(Clone, Default, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct FileAttributes {
    #[id(0)]
    pub content: bool,
    #[id(1)]
    pub aux_data: bool,
}

#[auto_wire]
#[derive(Clone, Default, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct FileSpec {
    #[id(0)]
    pub key: Key,
    #[id(1)]
    pub attrs: FileAttributes,
}

#[auto_wire]
#[derive(Clone, Default, Debug, Eq, PartialEq, Serialize)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct FileRequest {
    // #[id(0)] # deprecated
    #[id(1)]
    pub reqs: Vec<FileSpec>,
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for FileContent {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        let bytes: Vec<u8> = Arbitrary::arbitrary(g);
        Self {
            hg_file_blob: Bytes::from(bytes),
            metadata: Arbitrary::arbitrary(g),
        }
    }
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for FileAuxData {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        let bytes: Vec<u8> = Arbitrary::arbitrary(g);
        Self {
            total_size: Arbitrary::arbitrary(g),
            sha1: Arbitrary::arbitrary(g),
            blake3: Arbitrary::arbitrary(g),
            file_header_metadata: Some(Bytes::from(bytes)),
        }
    }
}

#[auto_wire]
#[derive(Clone, Default, Serialize, Deserialize, Debug, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct HgFilenodeData {
    #[id(0)]
    pub node_id: HgId,
    #[id(1)]
    pub parents: Parents,
    #[id(2)]
    pub file_content_upload_token: UploadToken,
    #[id(3)]
    pub metadata: Vec<u8>,
}

#[auto_wire]
#[derive(Clone, Default, Debug, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct UploadHgFilenodeRequest {
    #[id(0)]
    pub data: HgFilenodeData,
}

#[auto_wire]
#[derive(Clone, Default, Serialize, Deserialize, Debug, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct UploadTokensResponse {
    #[id(2)]
    pub token: UploadToken,
}
