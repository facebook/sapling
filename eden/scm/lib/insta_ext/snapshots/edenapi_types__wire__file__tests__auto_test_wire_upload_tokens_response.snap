---
source: file.rs
expression: "WireUploadTokensResponse :: arbitrary(& mut g)"
---
{"2":{"1":{"1":{"5":[154,77,0,155,19,212,76,218,174,127,211,79,1,180,239,159,71,129,237,47,133,207,59,183,155,157,133,85,4,46,38,85]}},"2":{"1":[36,58]}}}
